### PR TITLE
XSOAR CI/CD - Add pull request automation feature

### PR DESCRIPTION
**Pull request by adi88d**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-3028)

## Description
Added the pull request creation feature to `XSOAR CI/CD` pack to allow users create and update pull request from content zip file using Github integration.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/ContentManagement/README.md [hunk 1]
```diff
@@ -11,4 +11,4 @@ As part of this pack, you will also get out-of-the-box incident type, fields and
 
 For more information, visit our [Cortex XSOAR Developer Docs](https://xsoar.pan.dev/docs/reference/packs/content-management)
 
-![playbook.png](https://raw.githubusercontent.com/demisto/content/master/Packs/ContentManagement/docs-files/playbook.png)
+![playbook.png](Doc_files/playbook.png)
```
Comment (yaakovi) [inline Packs/ContentManagement/README.md:14]: don't use relative path in pack README.

File: Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+    "associatedToAll": false,
+    "associatedTypes": [
+        "Pull Request Creation"
```
Comment (yaakovi) [inline Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json:4]: what about the update?

File: Packs/ContentManagement/IncidentFields/incidentfield-PullRequestReview.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+    "associatedToAll": false,
+    "associatedTypes": [
+        "Pull Request Creation",
+        "Pull Request Creation"
```
Comment (yaakovi) [inline Packs/ContentManagement/IncidentFields/incidentfield-PullRequestReview.json:5]:
```suggestion
        "Pull Request Update"
```

File: Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    "closeForm": false,
+    "content": true,
+    "description": "Github username to set as pull request reviewer.",
+    "editForm": true,
+    "group": 0,
+    "hidden": false,
+    "id": "incident_cicdreviewer",
+    "isReadOnly": false,
+    "locked": false,
+    "name": "CI/CD Pull Request Reviewer",
```
Comment (yaakovi) [inline Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json:17]: what's the difference between this field and `Packs/ContentManagement/IncidentFields/incidentfield-PullRequestReview.json`

**Review by yaakovi — COMMENTED** [review]
regenerate the playbook images after the doc-review

File: Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json (hunk 4 above)
Comment (adi88d) [reply Packs/ContentManagement/IncidentFields/incidentfield-Reviewer.json:17]: CI/CD Pull Request Reviewer - its the user that review the PR
CI/CD Pull Request Review - the review for the PR for users that works on XSOAR and not familiar with github

File: Packs/ContentManagement/Playbooks/playbook-Pull_Request_Creation.yml [hunk 5]
```diff
@@ -0,0 +1,1406 @@
+id: Pull Request Creation
```
Comment (yaakovi) [inline Packs/ContentManagement/Playbooks/playbook-Pull_Request_Creation.yml:1]: please follow https://xsoar.pan.dev/docs/playbooks/playbook-conventions

File: Packs/ContentManagement/Scripts/SuggestBranchName/SuggestBranchName.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        return_results(CommandResults(
+            readable_output=branch_name,
+            outputs_prefix='AvailableBranch',
+            outputs=branch_name
+        ))
+
+    except Exception as ex:
+        demisto.error(str(ex))  # print the traceback
+        return_error(f'Failed to execute script. Error: {str(traceback.format_exc())}')
```
Comment (yaakovi) [inline Packs/ContentManagement/Scripts/SuggestBranchName/SuggestBranchName.py:38]:
```suggestion
        return_error(f'Failed to execute script. Error: {ex}', error=ex)
```

File: Packs/ContentManagement/Scripts/GetPrBranches/GetPrBranches.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+            branch = incident.get('CustomFields', {}).get('cicdbranch')
+            if branch:
+                branches.append(branch)
+
+        output = {"hidden": False, "options": sorted(branches)}
+        return_results(output)
+
+    except Exception as ex:
+        demisto.error(str(ex))  # print the traceback
+        return_error(f'Failed to execute script. Error: {str(traceback.format_exc())}')
```
Comment (yaakovi) [inline Packs/ContentManagement/Scripts/GetPrBranches/GetPrBranches.py:25]:
```suggestion
        return_error(f'Failed to execute script. Error: {ex}', error=ex)
```

**Review by yaakovi — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@yaakovi Doc review completed.

**Comment by glicht** [comment]
@idovandijk this PR is waiting for your review

**Comment by idovandijk** [comment]
> @idovandijk this PR is waiting for your review

Been working on many contributions that had accumulated, I will get to it shortly.

**Review by idovandijk — COMMENTED** [review]
@adi88d Impressive work, as expected of you. You've put a lot of effort into this PR, I want to commend you for that. In reviewing this I checked the validity and connectivity of the different content items and focused on the logic, the user experience and the modularity and generic aspect of the PR. I like to number my points so you can easily refer to any of them by the number:

1. Why is it that when the user contributes a number of items at once, they are committed separately? It is better if we can change that to be a single commit per contribution creation/update.
image

2. It is better to make the template an input of the `CommitFiles` script rather than hardcoded:
image
it may pose a small challenge but it should be possible, I believe we've done it in the IAM pack.

3. In the incident layout, consider reorganizing the fields so that they're fully visible. Also the comment might be a longer field. Maybe change the fields to cards:
image

4. See if you can improve this part that we've talked about in the demo as well:
image

The fact that we have 2 fields that do the same thing may be confusing despite the tooltips. Maybe we can fetch them from the repo? Maybe we can have the first option as "Other branch name..." and when that is selected we can display the 2nd textbox which allows shorttext input?

5. Let's explain the logic behind the generation of the branch name if nothing is specified:
image

6. Maybe this field can fetch the names from the repo too?
image

7. When we choose the `Pull Request Update` incident type, how does it recognize which files to update? Does it happen by filename of content item ID or item Name? What happens if we want to change the name of a content item?

8. Not sure if possible because it has DT - but see if this message can also be configurable as a playbook input:
image

9. In terms of looking at future updates - I think that unless "Pull Request Create" and "Pull Request Update" will be using the same playbook both for **creation** and **update** of both **Github** and **Gitlab** and other products, we need to think how to make it modular. I think that we need to split the playbook into different paths. First the common tasks (getting the zip, extracting the files), then 2 different branches depending on which integration they choose to use (Github/Gitlab). The choice should be via playbook input.
Preferably there will be a subplaybook for Github and in the future we can add a subplaybook for Gitlab.
Also, check how Microsoft Team works. As we will probably be adding it in the future, it may be better to just provide a "ChannelName" rather than "SlackChannel" input - so we can use the same input for both (again, not sure if Teams uses "channels", worth checking).

10. The whole part which relates to the notification should be under a "Notification" section. See if it's possible to make the `SlackMessageText` key unrelated to Slack, so it could be a MessageText key which will be used regardless if they use Slack or Teams (or email). Referring to these tasks:
image

11. Note that the pack name is missing from the "update" flow in task 35:
image
VS
image

I see that it should exist in context in the `Pull Request Update` flow so let's add it to the message.

12. The repository name should also be a playbook input, because the repo configured in the integration is only used in the fetch process, but in the commands it needs to be specified from what I've experienced with the Github integration:
image

13. Finally please try to make the task names a bit more meaningful. Example:
image

the left one creates a PR with the title as the pack name. Please also add descriptions to all the tasks. The SDK validation should fail without it but I think something is wrong with the validation lately.

I wrote many points but some are just suggestions. I would focus mostly on the points for making this generic/modular or the quick wins.

**Comment by adi88d** [comment]
Thanks @idovandijk for the review:
1. its not possible due to limitation of github API: see the documentation [here](https://docs.github.com/en/rest/repos/contents#create-or-update-file-contents)
2. I will change the template to be the input of the script.
3. will do
4. fetch the branch name from the repo is not an option because we dont know which branches created from XSOAR, im checking about update the second branch name field when changing the first one.
5. i will update the description for this field.
6. its not possible - we dont have a command for getting the users list and also the result might be a long and not useful.
7. when running the playbook on `pull request update` incident type, the script look for the same file name in the same path - if the file exists it will update it, when changing the name of the file the script treat it like a new content item.
8. I will try it and see if its possible.
9. we will redesign the playbooks according to your inputs.
10. will do.
11. will update the message template.
12. will add the repo name to playbook inputs.
13. will update the tasks name.

**Review by idovandijk — APPROVED** [review]
Following the demo - awesome.
Please incorporate the 2 small changes (rename main pb to have `- Generic` suffix, and rename task that sends noitification (it's not just for Slack).
Feel free to merge after that.

**Pull request merged**